utils/netlist-checks/io-checks.stanza

```diff
@@ -0,0 +1,508 @@
+          % [ref $ pin(open-collector), iol(driver), ipk],
+        locators =             [pin(open-collector)])
+  
+
+pcb-check input-and-generic-pin-max-voltage-rating (net-voltage:Toleranced, pin-categories:DigitalPinCategories) :
+  ; Now we need to check on any io-pins connected to this output pin
+  val input-pins = cat(ios(pin-categories), inputs(pin-categories))
+  for p in input-pins do :
+    val pin = pin(p)
+    val input-prop = prop(p)
+    ; Check that the min net voltage is above the rating of the input pin
+    match(input-prop:DigitalInput|DigitalIO) :
+      val vih-v = vih(input-prop)
+      #CHECK(condition = min-value(net-voltage) > vih-v,
+             name = NAME
+             description = DESCRIPTION
+             category = CATEGORY
+             subcheck-description = "Check the net with a net-voltage property is higher than the vih pin specification",
+             pass-message = "%_'s vih specification (%_V) is lower than the net voltage (%_V)"    
+              % [context(pin), vih-v, net-voltage],
+             fail-message =  "%_'s vih specification (%_V) is greater than the net voltage (%_V)" 
+              % [context(pin), vih-v, net-voltage],
+             locators = [pin])
+
+
+; Returns a list of pullup resistors and their opposing net voltages
+defn pullup-resistors (open-collectors:Tuple<DigitalPin>,
+                       pin-categories:DigitalPinCategories) :
+  val collected-net-voltages = Vector<Toleranced>()
+
+  defn is-resistor? (r) :
+    has-property?(r.resistor) and
+    has-property?(r.resistance)
+
+  defn collect-net-voltages-and-returns-if-any? (instance:Instance) -> True|False :
+    val voltages = 
+      for p in pins(instance) seq? :
+        property?(p.net-voltage)
+    val result = not empty?(voltages)
+    add-all(collected-net-voltages, voltages as Seq<Toleranced>)
+    result
+
+  ; lazy 'and', connected-to-net? can be heavy
+  defn pull-up? (instance) :
+    is-resistor?(instance) and
+    collect-net-voltages-and-returns-if-any?(instance)
+
+  ; First, we search for the pull-up-resistors that might be connected to this pin
+  val pull-up-resistors = all(pin-categories) $> map{containing-instance!, _}
+                                              $> unique
+                                              $> filter{pull-up?, _}
+                                              $> to-tuple
+  
+  [pull-up-resistors, to-tuple(collected-net-voltages)]
+
+; Compute the parallel resistance of a tuple of resistors. 
+; Note on numerical stability : 
+; 
+; In the empty array case, we return false (instead of +INFINITY). 
+; Negative resistances are forbidden and will cause a fatal error.
+; 
+; Zero-ohm resistors in parallel will result in 0 total resistance.
+;
+defn parallel-resistance (resistors:Tuple<Instance>) -> Double :
+  val resistances = map({property(_.resistance) as Double}, resistors)
```

Comment:
`resistance` property guaranteed by prior call to `pullup-resistors`, nice!

Reply:
Which is why this helper should stay private...